lua/kznllm-v3/presets/basic.lua

```diff
@@ -0,0 +1,252 @@
+local utils = require 'kznllm-v3.utils'
+local buffer_manager = (require 'kznllm-v3.buffer').buffer_manager
+local Path = require 'plenary.path'
+local api = vim.api
+
+local anthropic = require 'kznllm-v3.specs.anthropic'
+local openai = require 'kznllm-v3.specs.openai'
+
+local M = {}
+
+---@class BasicPreset
+---@field id? string
+---@field description? string
+---@field invoke fun(opts: table)
+---
+-- Preset configuration builder
+---@class BasicPresetConfig
+---@field id string
+---@field description string
+---@field curl_options BaseProviderCurlOptions
+
+---@class BasicPresetBuilder
+---@field provider BaseProvider
+---@field template_path Path
+local BasicPresetBuilder = {}
+
+---@param config { provider: BaseProvider, template_path: Path }
+---@return BasicPresetBuilder
+function BasicPresetBuilder:new(config)
+  local instance = { provider = config.provider }
+  setmetatable(instance, { __index = self })
+  return instance
+end
+
+function BasicPresetBuilder:make_data(curl_options, prompt_args)
+  error("make_data NOT IMPLEMENTED", 1)
+end
+
+---@param config BasicPresetConfig
+---@return BasicPreset
+function BasicPresetBuilder:build(config)
+  return {
+    id = config.id,
+    description = config.description,
+    invoke = function(opts)
+      local provider = self.provider
+
+      local user_query = utils.get_user_input()
+      if user_query == nil then return end
+
+      local selection = utils.get_visual_selection(opts)
+
+      local current_buf_id = api.nvim_get_current_buf()
+      local current_buffer_context = buffer_manager:get_buffer_context(current_buf_id)
+
+      local prompt_args = {
+        user_query = user_query,
+        selection = selection,
+        current_buffer_context = current_buffer_context,
+        replace = not (api.nvim_get_mode().mode == 'n'),
+        context_files = utils.get_project_files({
+          stop_dir = Path:new(vim.fn.expand '~'),
+          context_dir_id = '.kzn'
+        }),
+      }
+
+      self:make_data(config.curl_options, prompt_args)
+
+      if opts.debug then
+        local scratch_buf_id = buffer_manager:create_scratch_buffer()
+        local debug_data = provider:make_prompt_from_template({ filename = 'debug.xml.jinja', prompt_args = config.curl_options.data })
+
+        buffer_manager:write_content(debug_data, scratch_buf_id)
+        vim.cmd 'normal! Gzz'
+      end
+
+      local args = provider:make_curl_args(config.curl_options)
+      local job = buffer_manager:create_streaming_job(provider, args)
+      job:start()
+    end
+  }
+end
```

Comment:
I think this is easier to understand than the thing before - no extmarks being passed around or anything and it's all tucked behind buffer manager.

The template differences between anthropic vs. openai spec is more nicely abstracted away

```lua
local AnthropicPresetBuilder = BasicPresetBuilder:new({
  provider = anthropic.AnthropicProvider,
})

function AnthropicPresetBuilder:make_data(...)
...

---@class OpenAIPresetBuilder : BasicPresetBuilder
local OpenAIPresetBuilder = BasicPresetBuilder:new({
  provider = openai.OpenAIProvider,
})
function OpenAIPresetBuilder:make_data(...)
...
```